Hare 3 initial implementation

piersy opened the pull request:
This PR is intended to give a look into the current state of the hare3 implementation (it is not functional).

There is a document [description.md](https://github.com/spacemeshos/go-spacemesh/blob/72e20642fea9939d4b0f7765f1430719843b692c/hare3/description.md) that provides a high level description of the protocol and its parts.

There is also the current state of the protocol implementation under `./hare3`.

piersy committed:
Add hare3 outline

piersy committed:
Adjust structure of protocol and handler

The broker needs to know from the handler whether it should relay a
message to the network. Having that communication happen through a
channel seems over complicated, so instead the broker will call handle
directly and the handler and the protocol will use a mutex to
synchronise their actions. Neither the handler or protocol contain any
blocking code (except for the lock) and the code is not performing
costly transactions so this shouldn't pose a problem to the broker by
causing the broker to wait.

piersy committed:
Switch from using hare3.Hash20 to types.Hash20

piersy committed:
Thinking about the path to take

I think we can run everything (including early messages) through the
handler and then construct the actual protocol bit later, because really
the handler is just a bunch of accumulators.

The set Si can be provided to the protocol when we start to run it, it
doesn't need to be given at construction time.

piersy committed:
Wip

have decided that we process early messages fully through the hare3 handler
and that that means that we need to build the hare3 handler in the
broker in response to early messages.

That means that we want to return the gradecaster and thresh graded
gossiper upon consensus process registration so that those can be built
into the consnesus process.

This may not be an ideal final design since the contents of hare/hare.go
are looking less and less useful, but for now we'll do this to get the
broker integration out of the way.

It feels to me that ultimately the hare and broker will merge to become
one thing and a lot of the hare functionality will be repackaged into
smaller mor targetted services.

Next step right now is to split the handler object defined in the broker
to be a message store and a separate handler.

piersy committed:
Split handlers from message store and return handler

piersy committed:
Cleaned up broker

Could now produce diagram explaining broker and hare integration.

piersy committed:
Add early message handling

piersy committed:
wip

piersy committed:
Fixup comments

piersy committed:
Re-organise broker

piersy committed:
Update hare3.Handler to pass malfeasant messages to graded gossiper

piersy committed:
Add high level description of hare3 protocol

piersy committed:
Move new broker under hare3

piersy committed:
Re-instate old broker

piersy committed:
Move broker to own package

No need to contaiminate the algorithm's package with all the broker
dependencies.

piersy committed:
wip

piersy committed:
Remove unused parameter in broker

piersy committed:
Musings on weak coin and running

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/hare3-new

piersy committed:
Wip

piersy committed:
Wip is it done?

piersy committed:
Remove synced check on broker handle

This would have caused problems with early messages, since by definition
we would not have been synced with the next layer.

piersy committed:
Got weak coin handling in place

piersy committed:
Fix lint error

piersy committed:
Lint fixes

piersy committed:
Little clean up and spelling fixes

piersy committed:
Clean broker a bit

piersy committed:
Fix LayerClock interface

piersy committed:
Add round provider

The round provider allows the sub-protocols to understand what is the
current round, which they require in order to correctly process
messages.

piersy committed:
Implement TrhesholdGradedGossiper

piersy committed:
Add test for DefaultThresholdGradedGossiper

piersy committed:
Complete testing of DefaultThresholdGradedGossiper

piersy committed:
Extend documentation for GradedGossiper

piersy committed:
Add gradecaster implementation

Update docs a bit

piersy committed:
Implement GradedGossiper

piersy committed:
Move graded gossiper to own file

piersy committed:
Move ThresholdGradedGossiper to own file

piersy committed:
Move Gradecaster to own file

piersy committed:
Move ThresholdGradedGossiper test to own file

piersy committed:
Rename Gradecast file

piersy committed:
wip

piersy committed:
Extract common testing values

piersy committed:
Update use of gradecast

piersy committed:
Add test of gradecast

piersy committed:
Add more test cases for gradecast

piersy committed:
A little cleanup in algorithm.go

Also a fix for early messages

piersy committed:
Skip handling duplicate messages in broker

Also remove unnecessary message handling for early messages.

piersy committed:
Broker update round to be of type AbsRound

piersy committed:
Add test for graded gossip

dshulyak commented on hare3/gradecast.go:
```diff
@@ -0,0 +1,126 @@
+// msgRound. Messages received too late or with insufficient grades will not be
+// retrievable. Subsequent malicious messages can also render previously good
+// messages un-retrievable.
+type Gradecaster interface {
```
what is the point of it being an interface? does it simplify testing? i would rather not mock this stuff

piersy replied:
Initially I started with an interface because I didn't have the implementation, I just added an implementation yesterday, and was considering whether the interface was still useful. I was planning to leave the interfaces till I'd finished all the testing in case a situation came up where we needed to mock something. I can't see that we would need to mock anything but it's no more work to remove the interfaces later rather than now.

dshulyak commented on hare3/graded_gossip.go:
```diff
@@ -0,0 +1,97 @@
+// GradedGossiper works as a filter, for the given message inputs it returns
+// one of 3 results indicating what action to take with the message it was
+// invoked for.
+type GradedGossiper interface {
```
what the point of it being an interface? does it simplify testing?

piersy replied:
Same answer as above

dshulyak commented on hare3/algorithm.go:
```diff
@@ -0,0 +1,478 @@
+	tgg ThresholdGradedGossiper
+	gc  Gradecaster
+	lc  LeaderChecker
+	mu  *sync.Mutex
```
please put it on top of stuff that is synchronized in final version

piersy replied:
Do you mean the following ?

```
mu  *sync.Mutex
x   someType // synchronised by mu

y   someOtherType // not synchronised by mu
```
Is this a convention that we have, or in general a convention? I've never heard of this before.

dshulyak replied:
yes like that, it is a convention atleast in spacemesh codebase, and i definitely stumbled on it before, though i will need to find sources if you need them

another approach that i like is to nest sychronized fields after mutex, it is general practice in cockroachdb codebase

```
mu struct {
    sync.Mutex
    x someType
}
```

piersy replied:
The nesting is nice, more explicit and less likely to get mangled by future updates

piersy replied:
I'd be in favour of nesting, unless you'd rather stick with the line locality approach?

dshulyak reviewed (commented):
one thing that will be very important to redefine for initial integration are messages, they should be encoded as a scale enum where every specific message doesn't carry misleading junk (similar to protobuf oneof), and there should be no recursive types in them

there is an example with malfeasance proof where message basically encoded as `byte || scale type ` , and decoded then checks `byte` value before choosing how to decode.

dshulyak commented on hare3/algorithm.go:
```diff
@@ -0,0 +1,478 @@
+//
+// from the abs round both round and message type can be
+// the iteration and the round within that iteration can be derived.
+type AbsRound int8
```
this needs to be larger (max number of iterations is 40) and unsigned if it is going to be encoded with scale. note that we use compact encoding with scale, where integer is encoded into 1, 2, 4, 9 bytes depending on its size

piersy replied:
Thank you, I was wondering if this would be sufficient

dshulyak commented on hare3/graded_gossip_test.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+func TestGradedGossip(t *testing.T) {
+	gg := NewDefaultGradedGossiper()
```
please use table tests, or named subtests with extracted function like you mentioned previously

piersy replied:
The reason I didn't do this here is I think its quite clear with the comment what is happening at each point, and when the test fails you will get a line number taking you to the exact point of the failure. I think the comments are too long to serve as good subtest names, so by adding subtests you'll end up with lots of tests with largely meaningless names which are adding boilerplate to the test. I don't really see what you'll gain from doing that?

dshulyak replied:
it allows to debug a single failed test instead of all. for example if can add breakpoint or print statement inside a function that is tested and run single test, and it will trigger only in that function

tbh i don't see downsides of doing subtests, even if you leave name field empty

dshulyak commented on hare3/algorithm_test.go:
```diff
@@ -0,0 +1,36 @@
+	r1   = NewAbsRound(0, 1)
+	r2   = NewAbsRound(0, 2)
+	r3   = NewAbsRound(0, 3)
+	r4   = NewAbsRound(0, 4)
```
let's avoid global state for tests. if you want to reuse same setup for tests that can be accomplished with simple object

piersy replied:
I not sure I would consider this global state, since these are not intended to be written, just read. Unfortunately due to the types involved they can't be made consts. Could you elaborate on your suggested sue of a simple object, I'm not sure exactly what you are suggesting?

dshulyak replied:
really poor implementation of this pattern is in spacemesh api modules. in your case it is smaller but it is still not good.
for example if i need more than 3 ids, will i add more here? same for values1/2/3

i think better pattern is to write stateful tester class that will generate and manage collections of objects. so for example in this simple case it would be

```go
type hareTester struct {
    ids []int
    values []types.Hash20
}

func (t hareTester) getValues(positions ...int) []types.Hash32 {
    // and here it just selects values using positions, maybe even extends values is position is not found
}
```

piersy replied:
Hi @dshulyak sorry for the delay in getting back to you here. Is your concern that we have global state or is it that we'll need to keep adding values?

dshulyak replied:
the pattern with global state setup is a pain for next person that will work on this module

piersy replied:
I don't see why that is.

IIUC what you are proposing is that we build some object that can then be used to generate test values for each test.

I see that as more of a pain because every test then needs to define the test values it needs and likely the naming will not be consistent between tests so it becomes harder to read all the tests.

So you'll end up with more code required for each test, it will take longer to read all the tests, because there won't be consistency between the variables used in each test and also you need to actually write the code for the value generator object.

Am I misunderstanding your suggestion, is there something I'm not considering?

dshulyak replied:
so you are saying that you will be able to test everything with 3 ids and 3 different set of values? can you check API module, does testing there look ok to you?

piersy committed:
Separate tests into subtests

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/hare3-new

piersy committed:
Update broker use cdb for malfesance access

Instead of mesh, this was changed on develop.

piersy committed:
Export NewEligibilityValidator

So it can be accessed outside of hare.

piersy committed:
Add implementation for LeaderChecker

Modify exising protocol to handle types.NodeID

piersy committed:
Add network gossiper implementation

piersy committed:
Remove malfeasance handlers dependency on consensus process

piersy committed:
Replace old hare with new

piersy committed:
Remove now invalid tests

Since we are not checking eligibility as part of the malfeasance these
test are now invalid.

piersy committed:
Export fields of OutputMessage

piersy committed:
Add message builder

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/hare3-new

piersy committed:
Test kinda working, atxs building beacon being calculated

Hare is not currently working though

piersy committed:
Implement proposal set hash calculation

piersy committed:
Ensure that broker builds handler

piersy committed:
Correct usage of Vi

We were not initialising it before trying to access it and were also
accessing incorrect indexes. We were using the grade as the index
instead of calculating the index as 5-grade

piersy committed:
Remove debug additions

piersy committed:
Correct threshold calculation in threshold graded gossiper

piersy committed:
Ensure that message builder signs messages

piersy committed:
Add log to hare3 protocol runner

piersy committed:
Extract interface for round provider

piersy committed:
Rename gradecaster constructor

piersy committed:
Ensure we handle sorted sets of values for easy comparison

piersy committed:
Clarify comment in broker

piersy committed:
Don't send notification msg if no candidate value

piersy committed:
Add unit test of consensus

dshulyak commented on hare3/algorithm.go:
```diff
@@ -0,0 +1,502 @@
+	// Each index "i" holds a map of sets of values from valid proposals
+	// indexed by their hash received in iteration "i"
+	Ti  []map[types.Hash20][]types.Hash20
+	Si  []types.Hash20
```
could you help me understand why protocol needs to store history for all previous iterations? and not two at most

piersy replied:
That's what is specified in the whitepaper, at several points the protocol inspects the whole history of valid proposals to check whether the message it's considering is voting for a valid proposal. As to why the protocol inspects the whole history and not just the last 2 iterations I'd suggest asking Tal.

dshulyak replied:
i think it makes more sense if hardLocked and locked are single variables which are reset when softLock / hardLock rounds start

do you see a reason why Ti is not just a map[types.Hash20][]types.Hash20 ?when you run softLock / hardLock there should be only values from prev iterations

dshulyak commented on hare3/algorithm.go:
```diff
@@ -0,0 +1,502 @@
+	Li []*types.Hash20
+	// Each index "i" holds values considered valid up to round "i+1"
+	// TODO convert this to an object that accepts grades, so that it can be more closely mapped to the paper
+	Vi [][]types.Hash20
```
i is not an iteration here?

dshulyak commented on hare3/algorithm.go:
```diff
@@ -0,0 +1,502 @@
+		}
+		// If we didn't get a set from threshold gossip then use our local candidate
+		if set == nil {
+			set = p.Vi[1]
```
could you name this as some constant?

piersy replied:
Not sure what you mean here, also are you referring to p.Vi or set ?

The current names are not particularly intuitive, but they do follow the naming in the whitepaper which is helpful for cross checking the code against the whitepaper.

dshulyak replied:
i am just confused about `1`

piersy replied:
Oh, right. Yes I plan to wrap this in a type that makes it more clear, 1 here means the proposals that were received in round 1 or lower, which implies that the proposals have grade d-1 or higher (where d is the degree of the grading, and is 5). So 1 means proposals with grade 4 or 5.

dshulyak reviewed (commented):
types in hare3 is actually simpler than i thought, e.g. there is basically one type. it is still needs to be defined outside original hare package though

have you considered what is the hare3 bandwidth usage without gossip overhead? max 800 messages, 6 rounds, each message ~1KB (50 * 20), so - 5MB or 2.5MB depending if stop gossiping once threshold reached, am i right?

piersy committed:
Adjust TestReachingConsensus

piersy committed:
Correct algorithm object initialisation

Also ensure we don't send proposal messages without a value.

piersy committed:
Add test of not reaching consensus

dshulyak commented on hare3/algorithm.go:
```diff
@@ -0,0 +1,502 @@
+type AbsRound int8
+
+// NewAbsRound constructs a new AbsRound from the given iteration (j) and round (r).
+func NewAbsRound(j, r int8) AbsRound {
```
i think this type doesn't have an advantage over type where Round and Iteration are stored separately

piersy replied:
It's more compact, than storing both separately, its also easier to compare since a given abs round only has one representation, whereas if we stored round and iteration separately there can be multiple representations for a single AbsRound (E.G. {iteration:1, round:0} {iteration:0, round:7} ).

What would be the benefit of storing round and iteration separately?

dshulyak replied:
but we will need to fit atleast 40 iterations and max 6 rounds, so 2 bytes either way, right?

we know when to increment iteration and when to increment round, so i thought that it is simpler to have struct with two fields and avoid additional arithmetic

> {iteration:0, round:7}

this is invalid, right?

piersy replied:
>but we will need to fit atleast 40 iterations and max 6 rounds, so 2 bytes either way, right?

Yes 2 bytes either way, by compact I meant we could fit more iterations in if we used a single value, several thousand as opposed to 256, it doesn't seem particularly important to have that though.

>we know when to increment iteration and when to increment round, so i thought that it is simpler to have struct with two fields and avoid additional arithmetic

If you look at the implementation of `NextRound` it just increments the round, it doesn't need to know about iterations or rounds, so we don't really know when to increment iteration or round explicitly in the current implementation.

I feel like if we represented round and iteration separately we'd just end up moving the existing logic to handle updating and calculating the absolute round, but I don't think it would get simpler.

>>{iteration:0, round:7}
>this is invalid, right?

You could consider it invalid or you could consider it as iteration 1 round 0, if you consider it as invalid then that's going to  require some code to validate updates, and that implies errors and that introduces extra complexity.

piersy committed:
Change how rounds are tracked in hare3 protocol

Now the round in the hare 3 protocl will closely track the real round
timing. This will allow us to share exactly the same round instance
between the protocl and the handler, which just makes things simpler.

piersy committed:
Share round instance between handler and protocol

piersy committed:
Add test showing network of 2 reaching consensus

piersy commented:
> types in hare3 is actually simpler than i thought, e.g. there is basically one type. it is still needs to be defined outside original hare package though

What types are you referring to here?

> have you considered what is the hare3 bandwidth usage without gossip overhead? max 800 messages, 6 rounds, each message ~1KB (50 * 20), so - 5MB or 2.5MB depending if stop gossiping once threshold reached, am i right?

I haven't given it consideration yet.

To clarify though there are 7 rounds (8 if you include the preround)
But messages are only sent for the preround, proposal, commit and notify rounds.
The protocol needs at least 2 iterations to complete, and so that makes 7 rounds in total  (4 in iteration 0 and 3 in iteration 1) where nodes send a message.

So we'd be looking at 800 * 7 messages being received by each node for an instance of hare 3.

dshulyak commented:
> What types are you referring to here?

i meant messages that go over the wire. i misunderstood it yesterday, it looks like Prepare/Propose contains the slice, and Commit/Notify only a single hash, so my estimate was completely wrong

dshulyak commented on hare3/runner/runner.go:
```diff
@@ -0,0 +1,394 @@
+		}
+		select {
+		// We await the beginning of the round, which is achieved by calling AwaitEndOfRound with (round - 1).
+		case <-r.clock.AwaitEndOfRound(uint32(round - 1)):
```
this is another part where i am confused, based on my understanding protocol needs to wait for 5 network delays (in iteration 0, and then 4). according to this clock it looks like it will wait for soft and hard llock too

- preround -> WAIT -> propose (only in first)
- propose -> WAIT -> WAIT -> commit
- commit -> WAIT -> notify
- notify -> WAIT -> TERMINATED ?

piersy replied:
If you're referring to the note mentioning that round 1 (soft-lock) is not required in iteration 0,  I understood that as meaning it's not required to execute the code of round 1 as opposed to meaning that you go straight from round 0 to round 2 with only one WAIT.

The reason for my interpretation is that otherwise the specification of how valid values is calculated does not make sense, since they are intended to be calculated at the beginning of rounds 0, 1, 2 and 3.

dshulyak replied:
no, this is not what i was referring to. this code will run waits in the wrong places

dshulyak replied:
preround -> WAIT -> softlock -> WAIT -> hardlock -> WAIT propose

but it should be

preround -> WAIT -> softlock -> hardlock -> propose -> WAIT -> WAIT

piersy committed:
For now hare3 is not handling expected leaders

piersy committed:
Get correct atx when fetching proposals

Previously we were retrieving the atx published in the current epoch,
which is not actually active. Now we get the active atx published in the
previous epoch.

piersy committed:
Simplify testing, make all parties always eligible

piersy committed:
Correct the way that hare runner uses round

piersy committed:
Wait for wakeup delta before starting layer

Lifted round clock creation from runLayer to Run in order to wait in the
correct place.

piersy committed:
Clean up logging in broker

piersy committed:
Remove un-needed return statement

piersy committed:
Add more logging in runner

piersy committed:
Update node to be able to construct hare runner

piersy committed:
Introduce logs into hare algorithm

piersy committed:
Sync access to weak coin chooser

piersy committed:
Final part to reach consistent hare3 conversion

The leader selection was relying on the eligibility code which was
causing us sometimes to not have any leaders. For now hardcoded to make
everyone a leader.

piersy committed:
Update logging in hare3 algorithm

piersy committed:
log adjustements comments and tests

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/hare3-new

piersy committed:
Cleanup mod file

piersy committed:
Test is working and test is a test!

piersy committed:
Lint fixes

piersy committed:
Fix hanging test

Since node.App now waits for the grpc services to complete startup, we
need to ensure that they can startup by not trying to bind all nodes'
grpc services to the same ports.

piersy committed:
Re-instate module field in logs

It turns out removing this breaks some tests

piersy committed:
Introduce weight/eligibilitis into hare3

piersy committed:
Initial attempt to work eligibilities into hare3

piersy committed:
Merge remote-tracking branch 'origin/develop' into piersy/hare3-new

piersy committed:
Add job to run just problematic test

piersy committed:
Update committee params

piersy committed:
Correct setup

piersy committed:
Eligibility fixup

piersy committed:
Make unit tests job not failfast

piersy committed:
Try wget for downloads

piersy committed:
Skip TestSpacemeshApp_NodeService

It seems like its failing, but it could be a leftover goroutine from
another test that is failing.

piersy committed:
Revert "Skip TestSpacemeshApp_NodeService"

This reverts commit ab1ddfb3fcadcd2a697c38e34fa270743a4bb6dd.

piersy committed:
Skip TestConsensus

See if other node tests pass if this test is not run

dshulyak commented:
following changes are **required** to proceed with review, sorry for the list but it is easier for me this way

protocol changes:
- hare 3 should use new types, that encode preround and propose with a collection. commit, notify with a single hash
- hare 3 should output result for building block after receiving notify messages
- hard lock round should be skipped completely in the very first iteration. i was wrong about soft lock, so the skip is only for hard lock

other required changes:
- hare3 should use field based logger as any other module in the system. no strings formatting. also zap can be used directly, as this is where we are moving
- hare3 should not initialize logger based on the input from network
- duplicates from p2p layer are expected, and should not spam warnings
- no crash on database error, if you want exit protocol with error. what you did might be right approach so please create an issue to consider it later
- hardLock, softLock not a collection, map with values also not a collection, you can add iteration to each value in the map. i know that it is different from paper
- info logs should be limited to data when hare starts and terminates, right now it is spammy
- change should be limited only to hare3, if other changes are needed in other modules they should be in separate prs
- mutex should be a first field everywhere
- this code is not sufficiently tested, i suggest implementing generator for messages and testing that output matches expected after a sequence of messages was processed. i would like to have framework similar to tortoise (latest) but significantly simpler
- hare instance can be instantiated when layer starts, and 1st message is expected only after wakeup delta. with this assumption handling of early message is not required - if someone sends message before your layer started thats invalid. 

other asks:
- verbosity in the code can be lower, in some places code is simpler then the comments that describe it 
- code can be also more compact, i suggest limiting hare3 files to protocol.go that contains everything needed for clean state machine like you have. and hare.go that handles messages and routes them to the protocol
- interface for output hare result / coin using channels wasn't implemented
- metrics are needed

dshulyak commented:
we decided to go with alternative implementation plan
